### ISSUE TITLE
client compilation failed under official gcc 11.2.0 container with strict c++20 standard enabled

### ISSUE THREAD

**Issue by nkinani** [eclipse-cyclonedds/cyclonedds-cxx#181]
### Infos: 
- host : official [GCC 11.2.0](https://hub.docker.com/_/gcc/) devcontainer
- libs : 
    - Eclipse Iceoryx v1.0.1
    - Eclipse CycloneDDS v0.8.1
    - Eclipse CycloneDDS-CXX v0.8.1
 - misc libs dependencies : git, cmake, bison, openssl, ... 

### HowTo:

1. clone the repo & switch to tags/0.8.1
2. update the _[cyclonedds-cxx/exemple/helloworld/CMakeLists.txt:37,38](https://github.com/eclipse-cyclonedds/cyclonedds-cxx/blob/120c34915a129c1189f6834b967616c248355b8f/examples/helloworld/CMakeLists.txt#L37)_ to enforce strict c++20 as follow :
```cmake 
set_property(TARGET ddscxxHelloworldPublisher PROPERTY CXX_STANDARD 20 CMAKE_CXX_EXTENSIONS OFF CMAKE_CXX_STANDARD_REQUIRED ON)
set_property(TARGET ddscxxHelloworldSubscriber PROPERTY CXX_STANDARD 20 CMAKE_CXX_EXTENSIONS OFF CMAKE_CXX_STANDARD_REQUIRED ON)
```
3. under the build dir, run
```shell 
cmake -DCMAKE_CXX_FLAGS=-std=c++20 -DCMAKE_VERBOSE_MAKEFILE=ON -DBUILD_EXAMPLES=ON .. && cmake --build .
```
4. then, you should see these errors (x4)
> cyclonedds-cxx/src/ddscxx/include/dds/core/detail/ReferenceImpl.hpp:63:33: error: template-id not allowed for destructor
>    63 | dds::core::Reference<DELEGATE>::~Reference<DELEGATE>()
> [...]
> cyclonedds-cxx/src/ddscxx/include/dds/topic/detail/TTopicImpl.hpp:211:31: error: template-id not allowed for destructor
>   211 | dds::topic::detail::Topic<T>::~Topic<T>()
> [...]
> cyclonedds-cxx/src/ddscxx/include/dds/sub/detail/TDataReaderImpl.hpp:514:34: error: template-id not allowed for destructor
>   514 | dds::sub::detail::DataReader<T>::~DataReader<T>()
> [...]
> cyclonedds-cxx/src/ddscxx/include/dds/pub/detail/DataWriterImpl.hpp:403:34: error: template-id not allowed for destructor
>   403 | dds::pub::detail::DataWriter<T>::~DataWriter<T>()

### Reason to do so: 

Our client library who use cyclonedds-cxx is dependent of some c++20 features 

### Solving: 

Following [this solved issue](https://wg21.cmeerw.net/cwg/issue2237) and especially the _clause 15_, it lead me to the correction of each template-id destructor definition in the previous files by commenting them as follow : 
1. _[cyclonedds-cxx/src/ddscxx/include/dds/pub/detail/DataWriterImpl.hpp:403](https://github.com/eclipse-cyclonedds/cyclonedds-cxx/blob/120c34915a129c1189f6834b967616c248355b8f/src/ddscxx/include/dds/pub/detail/DataWriterImpl.hpp#L403)_
```c++
template <typename T>    
dds::pub::detail::DataWriter<T>::~DataWriter/*<T>*/()   
{
    /* some code eluded */
}
```
2. _[cyclonedds-cxx/src/ddscxx/include/dds/sub/detail/TDataReaderImpl.hpp:514](https://github.com/eclipse-cyclonedds/cyclonedds-cxx/blob/120c34915a129c1189f6834b967616c248355b8f/src/ddscxx/include/dds/sub/detail/TDataReaderImpl.hpp#L514)_
```c++
template <typename T>
dds::pub::detail::DataReader<T>::~DataReader/*<T>*/()
{
    /* some code eluded */
}
```
3. _[cyclonedds-cxx/src/ddscxx/include/dds/topic/detail/TTopicImpl.hpp:211](https://github.com/eclipse-cyclonedds/cyclonedds-cxx/blob/120c34915a129c1189f6834b967616c248355b8f/src/ddscxx/include/dds/topic/detail/TTopicImpl.hpp#L211)_
```c++
template <typename T>
dds::pub::detail::Topic<T>::~Topic/*<T>*/()
{
    /* some code eluded */
}
```
4. _[cyclonedds-cxx/src/ddscxx/include/dds/core/detail/ReferenceImpl.hpp:63](https://github.com/eclipse-cyclonedds/cyclonedds-cxx/blob/120c34915a129c1189f6834b967616c248355b8f/src/ddscxx/include/dds/core/detail/ReferenceImpl.hpp#L63)_
```c++
template <typename DELEGATE>
dds::pub::detail::Reference<DELEGATE>::~Reference/*<DELEGATE>*/()
{
}
```
Doing so will result in a correct compilation & execution.

### Discussion 

Is there any reason to keep these template-id definitions **_even without the strict c++20 support ?_** Could you explicit their purpose or need ? 

Thanks !

**Comment by k0ekk0ek** [comment]
Hi @nkinani, sorry for the late reply. I'm not really sure if there's a reason to keep these around. If there's no downside, then I'm inclined to include these changes, but I'll have to look into it myself too.

**Comment by k0ekk0ek** [comment]
Looks valid @nkinani. Verified your changes locally (with `STANDARD 11` too) and a bit of googling turned up a [StackOverflow](https://stackoverflow.com/questions/63513984/can-class-template-constructors-have-a-redundant-template-parameter-list-in-c2) post that mentions it's a mistake introduced in C++14. I think we can reasonably assume this change won't break builds on older compilers, so definitely in favor of integrating this. If you want/prefer, I think you should have the honor of issuing a PR, but I'm happy to create one if you prefer that.

**Comment by nkinani** [comment]
Hi @k0ekk0ek. Thanks for your response & your further investigation and I will happily let you issuing a PR about that little adjustment.

**Comment by k0ekk0ek** [comment]
Will do :+1: Thanks again for the report

**Comment by Vinpasso** [comment]
Hi everyone,

I'm also running into these compiler issues and my application will not build because of the compiler errors generated in the CycloneDDS-CXX headers, e.g.:

```
lib/cyclonedds/install/include/ddscxx/dds/pub/detail/DataWriterImpl.hpp:402:34:
error: template-id not allowed for destructor
```

I don't have any special settings in the CMake configuration file, except for `CMAKE_CXX_STANDARD 20`. The compiler is the ubuntu standard `gcc version 12.3.0`.

Are there any plans to fix this or is there a workaround?

Thank you for your time

**Comment by eboasson** [comment]
@Vinpasso, it is fixed in master #412. We have no plan for back porting it as it is only a compile-time warning.

I'll close this ticket now, it should've been back when we merged the PR. If the fix not being in 0.10.x is a problem for you then please open a new issue for that and we can discuss *that* issue.
